velox/functions/sparksql/tests/RegexFunctionsTest.cpp

```diff
@@ -111,5 +359,233 @@ TEST_F(RegexFunctionsTest, RegexMatchRegistration) {
+       {std::vector<std::string>{"short"}}},
+      positions);
+}
+TEST_F(RegexFunctionsTest, RegexReplaceDataframeNoMatches) {
+  // No matches
+  std::vector<int64_t> positions = {1, 1};
+  testingRegexReplaceRows(
+      {"apple", "banana"},
+      {"orange", "grape"},
+      {"fruit", "fruit"},
+      {{std::vector<std::string>{"apple"}},
+       {std::vector<std::string>{"banana"}}},
+      positions);
+}
+TEST_F(RegexFunctionsTest, RegexReplaceDataframeOffsetPosition) {
+  // Offset position
+  std::vector<int64_t> positions = {9, 6};
+  testingRegexReplaceRows(
+      {"apple pie apple", "grape banana grape"},
+      {"apple", "banana"},
+      {"fruit", "fruit"},
+      {{std::vector<std::string>{"apple pie fruit"}},
+       {std::vector<std::string>{"grape fruit grape"}}},
+      positions);
+}
+TEST_F(RegexFunctionsTest, RegexReplaceEmptyStringsAndPatterns) {
+  // Empty strings and patterns
+  std::vector<int64_t> positions = {1, 1};
+  testingRegexReplaceRows(
+      {"", "hello"},
+      {"", "llo"},
+      {"prefix ", " world"},
+      {{std::vector<std::string>{"prefix "}},
+       {std::vector<std::string>{"he world"}}},
+      positions);
+}
+TEST_F(RegexFunctionsTest, RegexReplaceDataframeCharacterTypes) {
+  // Multiple character types
+  std::vector<int64_t> positions = {1, 1};
+  testingRegexReplaceRows(
+      {"123ABC", "!@#"},
+      {R"(\d)", R"(\W)"},
+      {"X", "Y"},
+      {{std::vector<std::string>{"XXXABC"}}, {std::vector<std::string>{"YYY"}}},
+      positions);
+}
+TEST_F(RegexFunctionsTest, RegexReplaceDataframeBadPosition) {
+  // Larger offsets than string size
+  std::vector<int64_t> positions = {10, 15};
+  testingRegexReplaceRows(
+      {"apple", "banana"},
+      {"apple", "banana"},
+      {"fruit", "fruit"},
+      {{std::vector<std::string>{"apple"}},
+       {std::vector<std::string>{"banana"}}},
+      positions);
+}
+
+TEST_F(RegexFunctionsTest, RegexReplaceDataframeLastCharacter) {
+  std::vector<int64_t> positions = {5, 6};
+  testingRegexReplaceRows(
+      {"applez", "bananaz"},
+      {"z", "z"},
+      {"", ""},
+      {{std::vector<std::string>{"apple"}},
+       {std::vector<std::string>{"banana"}}},
+      positions);
+}
+
+// Test to match
+// https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala#L180-L184
+// This test is the crux of why regex_replace needed to support non-constant
+// parameters. Used position {0,0} out of convenience, ideally we create another
+// function that does not pass a position parameter.
+TEST_F(RegexFunctionsTest, RegexReplaceMatchSparkSqlTest) {
+  std::vector<int64_t> positions = {1, 1, 1};
+  testingRegexReplaceRows(
+      {"100-200", "100-200", "100-200"},
+      {"(\\d+)-(\\d+)", "(\\d+)-(\\d+)", "(\\d+)"},
+      {"300", "400", "400"},
+      {{std::vector<std::string>{"300"}},
+       {std::vector<std::string>{"400"}},
+       {std::vector<std::string>{"400-400"}}},
+      positions);
+}
```

Comment:
@mbasmanova I am still getting familiar with Velox as a codebase and this is my first PR so I could very easily be wrong, but from what I can tell it is a Vector function. 

These functions were modelled after the "class Re2SearchAndExtract final : public VectorFunction" and "class Re2Match final : public VectorFunction" in velox/functions/lib/Re2Functions.cpp, and I made an assumption that I would also have to create a VectorFunction. 

The constant parameters use case can be called on a with regex_replace('Hello World', 'l', 'L'), but it can also be called on a row with the same values. Additionally, if a user had a column of strings they wanted to do a batch of replacements on, they can do so with the constant pattern use case.

The non-constant parameters use case can be called with a rowVector like seen in the Spark UT test case, which attempts to directly match the Spark use case found here: https://github.com/apache/spark/blob/v3.2.2/sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala#L131 which operates on a dataframe. 

The testingRegexReplaceRows function converts the passed parameters into a rowVector, then the columns are passed into the function. 

I hope this answers your question, but to be honest I am not sure if it does. If this can be accomplished using a simple function I'd be glad to try that out, though I am still not super familiar with the everything the codebase has to offer. Another assumption I am making is that a simple function could only accomplish the regex_replace("Hello World", "l", "L") call for the constant parameters use case. Is that correct?  Or can they accomplish the same thing that the non-constant parameters use case is?

Apologies if I am misunderstanding anything or if I am being confusing. I would be glad to clarify anything.

Reply:
@codyschierbeck Thank you for the context. This is super helpful. Consider going over https://facebookincubator.github.io/velox/develop/scalar-functions.html and playing with simple function API. It should be possible to implement this function using that. The other functions you mentioned were implemented before simple function API became as powerful as it is now.

Reply:
@mbasmanova Hi Maria, I am trying to get more familiar with scalar functions vs vector functions and I have a question, if you don't mind. My basic understanding right now is that if compute/storage is required between rows, OR if it produces multiple values, a function must be a vector function. If all compute/storage for a row occurs within the context of that row AND it only produces a single value then it is better suited as a simple function. Is that correct? If not, what is a good way of deducing whether a function is better represented as either option?

In this specific example, during the non-constant pattern use case if we added an optimization to store compiled regexes in a map to reduce repeated compilation work, would that prevent this function from becoming a scalar/simple function? Another such potential optimization (though I'm unclear if the extra memory is worth the compute time decrease) can be found in the constant pattern use case. If we store "visited" strings in a map with their output after replacement as the value, we can reduce repeated work in that case as we only have to search for patterns in a given string once if there are duplicates. Would this also prevent a function from becoming a scalar/simple function? Or would we be able to store these maps as properties of the struct/class and reference that with each call?

Apologies if I am way off here. Just trying to wrap my head around the differences.

Reply:
@codyschierbeck Cody, thank you for looking into this.

All scalar functions produce only one value. Hence, regardless of whether function is implemented as simple or vector it produces single result value for each row of input.

Simple functions are stateful, e.g. you can use member variables to cache useful data to be re-used when processing different rows. See initialize() API for initializing state using constant inputs.

CC: @laithsakka 

I suggest to prototype a simple function and share that code along with any issues / questions you encounter along the way.

Reply:
@mbasmanova Hi Maria, yesterday I implemented the simple function version of regex_replace for SparkSQL Velox. Thank you for pointing me in the right direction and clarifying the questions I had. You were correct, it was much better suited as a simple function and the implementation was much more straightforward after a bit of ramping.

I would appreciate it if you could take a look and provide any suggestions for change. From what I can see, all of the test cases are passing and we are only seeing issues with the fuzzer. 

Thank you and please let me know if there is anything you would like me to change.